### DISCUSSION TITLE
API Question

### DISCUSSION THREAD

**Discussion by retrogamer1337** [benjaminjonard/koillection#1183] [Q&A]
Hi guys, I'm not a programmer and normally I play with chatGPT till something works as I want it to work. But for some reason I don't know how to start this.

In my Koillection used for videogames I have a collection called PS1. In this collection there are 250 Items. Every Item looks like that:

![image](https://github.com/user-attachments/assets/9a3ca5be-9d5e-4011-90cf-82883e19da9b)

Now I want to add a new Field to every link item called "PSX Datacenter" to make every item look like this in the end:

![image](https://github.com/user-attachments/assets/46b7395e-2ef2-4da6-a5e8-bbadbbf41601)

I have a file which contains a link for every "Playstation Catalog Number":

![image](https://github.com/user-attachments/assets/c55f5a0e-d071-435a-a033-bdc8b3f2f2b3)

So I would need to:

1. List all items and their corosponding field "Playstation Catalog Number".
2. Match the "PSX Datacenter" link to every "Playstation Catalog Number".
3. Create a new field "PSX Datacenter" for every item and insert the link there.

Can anyone help me on how to start? 

I would need to know:
1. How can I list all the content of "Playstation Catalog Number".
2. How can I create a new field named "PSX Datacenter" for each items of PS1.
3. How can I post a link to each newly created field "PSX Datacenter".

Thank you very much in advance.

**Comment by retrogamer1337** [comment]
I made it with thehelp of chatGPT ...

you need to get a list of all items of the collection (did this by scraping the collections webpage sourcecode) than you can add the link to every item by POST to URL/API/DATA like this:

{
    "item": "/api/items/{ITEM ID OF GAME}",
    "type": "link",
    "label": "PSX Datacenter",
    "value": "http://psxdatacenter-Link...",
    "position": 11,
    "visibility": "public"
}

**Comment by retrogamer1337** [comment]
So I managed to get everything done I wanted to with the api...

**1. create n new items.
2. add n data fields to each item and fill them.
3. upload a cover picture for each item.
4. add n more empty images to each item. 
5. upload an image file for each of the images created.**

I found out that I needed to use the following api requests for that:

1. POST /api/items 
provide existing collection and a new name for the item

2. POST /api/data
provide existing itemid and any number of new fields, each with type, label and value

3. POST /api/items/{itemID}/image
only the image file needs to be provided (content-type: multipart/form-data)

4. again POST /api/data (you may can also do this in step 2)
existing itemID and any number of new fields, each with type = image and any label

5. POST /api/data/{imageitemID}/image
(ATTENTION: I have called the itemID mentioned here imageitemID, because it is not the itemID of the main item, but that of the newly created empty image)

Intermediate steps were necessary to obtain the necessary IDs of the newly created items after the respective steps:

After you have created the new items in step 1, you can display the created IDs as follows: 
GET /api/collections/{collectionID}/items?page=1 
(you have to go through pages 1,2,3... until you have all the items)

After step 4, you can display the imageitemIDs of the newly created empty images as follows:
GET /api/items/{itemID}/data
All data records of the item are then listed, including the created images that do not yet contain any image files. however, these have their own itemimageID

In this way, I was able to add nearly thousand games from my game collection to Koillection in just a few hours.

I wrote the scripts for the automated api batch calls with powershell and had ChatGPT write json files based on my existing collection data to be able to import all the data. 
If anyone has any questions or needs help please contact me here. 

Now my games all look kind of like this:
![image](https://github.com/user-attachments/assets/051756f6-8248-44c7-9dc1-e6140d5f2358)

**Comment by rmarkus** [comment]
Hey retrogamer, this was already very helpful. Could you explain on 

> POST /api/data
> provide existing itemid and any number of new fields, each with type, label and value
>
I can only send 1 field per post request. 

```
POST url/api/data 
with the body: 
{
"item": "/api/items/0195c7a4-a91c-73ff-8476-56ec08c73ffd",
"type": "text",
"label": "Autor",
"value": "Jane Austen",
"visibility": "public"
}
```

works. But how to I can't figure out how to add several fields with one request for one item.
Any help would be greatly appreciated.

**Reply by retrogamer1337** [reply]
I did everything with single calls in batch. Im not sure if you can do more than one field in a single call. I did it with the postman api tool, you can input a api call with tokens and provide a list of tokens in file to make the tool do batch calls for each item of the list. 

Later on I wrote my own powershell scripts but as I remember also all single calls.

**Reply by rmarkus** [reply]
ok thanks for that reply. I do now several calls to set all metadata. Luckily, it is all local for my project ;-).
